### PR TITLE
Codable search results

### PR DESCRIPTION
**Pull request by VladislavFitz**
Implemented brand new SearchResults & FacetResults as Codable structures

### CODE REVIEW

File: Sources/EitherSingleOrList.swift [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+            return (value as? CustomStringConvertible)?.description ?? "\(value)"
+            
+        case .list(let list):
+            return list.description
+        }
+    }
+    
+}
+
+extension EitherSingleOrList: Codable where T: Codable {
```
Comment (spinach) [inline Sources/EitherSingleOrList.swift]: Maybe I misunderstood, but did we want to change that to `SingleOrList`?

File: Sources/JSON.swift [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+//  Created by Vladislav Fitc on 04/03/2019.
+//  Copyright © 2019 Algolia. All rights reserved.
+//
+
+import Foundation
+
+public enum JSON: Equatable {
+    case string(String)
+    case number(Float)
+    case object([String:JSON])
```
Comment (spinach) [inline Sources/JSON.swift]: is it better to call that `dictionary` instead of `object`? WDYT?

File: Sources/EitherSingleOrList.swift (hunk 1 above)
Comment (VladislavFitz) [reply Sources/EitherSingleOrList.swift]: Yes, in the end we decided to keep EitherSingleOrList. But I'm not a big fan of it, can rename if you assume that it's better.

File: Sources/JSON.swift (hunk 2 above)
Comment (VladislavFitz) [reply Sources/JSON.swift]: Agree

File: Sources/EitherSingleOrList.swift (hunk 1 above)
Comment (spinach) [reply Sources/EitherSingleOrList.swift]: As this is not public, I'm really fine with keeping it as well. It's understandable, so I'm good with the name

File: Sources/SearchResultsV2.swift [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+            case facets
+            case pagesCount = "nbPages"
+            case hits
+            case hitsPerPage
+            case processingTimeMS
+            case query
+            case queryID
+            case areFacetsCountExhaustive = "exhaustiveFacetsCount"
+            case message
+            case queryAfterRemoval = "queryAfterRemoval"
```
Comment (spinach) [inline Sources/SearchResultsV2.swift]: The 2 are the same, we can remove the right side

File: Sources/SearchResultsV2.swift [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+            case aroundGeoLocation = "aroundLatLng"
+            case automaticRadius
+            case facetStats = "facets_stats"
+        }
+        
+        /// Hits.
```
Comment (spinach) [inline Sources/SearchResultsV2.swift]: You can specify in the comment that it's the hits for the specific page specified.

**Pull request closed without merging**